Fixes for project loading progress display

PhysSong opened the pull request:
This PR currently includes two commits: first one fixes #3611, second one is enhancement mentioned in same issue. I set base branch to stable-1.2 because of https://github.com/LMMS/lmms/issues/3611#issuecomment-311695108 by @tresf.
It is okay to decouple two commits if needed(but I don't think the decoupling is needed).

@liushuyu This PR contains a new string which needs translation. Once it is merged, please add it to i18n source string list. 
see [this](https://github.com/PhysSong/lmms/blob/aee0f2b3a240cea7961ff71b4891dcf29c387474/src/core/TrackContainer.cpp#L127).

zonkmachine commented on src/core/TrackContainer.cpp:
```diff
@@ -124,6 +122,9 @@ void TrackContainer::loadSettings( const QDomElement & _this )
 			!node.toElement().attribute( "metadata" ).toInt() )
 		{
+			QString trackName = node.toElement().hasAttribute( "name" ) ? node.toElement().attribute( "name" ) :
+															   node.firstChild().toElement().attribute( "name" );
```
125 and 126 too long lines. The [coding conventions](https://github.com/LMMS/lmms/wiki/Coding-conventions) suggestion for ternary operators is something like this:

```
QString trackName = node.toElement().hasAttribute( "name" )
            ? node.toElement().attribute( "name" )
            : node.firstChild().toElement().attribute( "name" );
```

zonkmachine commented on src/core/TrackContainer.cpp:
```diff
@@ -124,6 +122,9 @@ void TrackContainer::loadSettings( const QDomElement & _this )
 		{
+			QString trackName = node.toElement().hasAttribute( "name" ) ? node.toElement().attribute( "name" ) :
+															   node.firstChild().toElement().attribute( "name" );
+			pd->setLabelText( tr("Loading Track %1 (%2/Total %3)").arg( trackName ).arg( pd->value() ).arg( Engine::getSong()->getLoadingTrackCount() ) );
```
It crashes here on command line render. If you wrap it in an `if`, something like this:

```
if( gui )
{
   pd->setLabelText( tr("Loading Track %1 (%2/Total %3)").arg(
   	trackName ).arg( pd->value() ).arg(
   	Engine::getSong()->getLoadingTrackCount() ) )
}
```

that should fix it.

Also. The line is much too long and wraps badly. Please try and use more lines.

zonkmachine commented:
Backtrace for crash on command line render

```
(gdb) bt
#0  QProgressDialog::value (this=0x0) at dialogs/qprogressdialog.cpp:649
#1  0x000000000056882b in TrackContainer::loadSettings (this=0x237dfc0, _this=...) at /home/zonkmachine/builds/lmms/lmms/src/core/TrackContainer.cpp:127
#2  0x0000000000552c0a in SerializingObject::restoreState (this=0x237dfe0, element=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/SerializingObject.cpp:70
```

PhysSong commented:
@zonkmachine Okay. I'll change the code and test it again.

PhysSong commented:
`pd->setValue( start_val + _this.childNodes().count() );`([this line](https://github.com/PhysSong/lmms/blob/aee0f2b3a240cea7961ff71b4891dcf29c387474/src/core/TrackContainer.cpp#L135)) also generated bugs, so I deleted it.
@zonkmachine I fixed code now. Could you test this again?

zonkmachine commented on src/core/Song.cpp:
```diff
@@ -1054,6 +1054,23 @@ void Song::loadProject( const QString & fileName )
+	m_nLoadingTrack=0;
+	for( int i=0,n=tclist.count(); i<n; ++i ){
+		QDomNode nd=tclist.at(i).firstChild();
+		while(!nd.isNull()){
```
I think the general idea is to put the braces on the same column/tab although I'm not 100% sure about the convention around this.

```

( .. )
{
  this
}
```


```

( .. ){
  not this
}
```

zonkmachine commented on src/core/Song.cpp:
```diff
@@ -1054,6 +1054,23 @@ void Song::loadProject( const QString & fileName )
+			{
+				++m_nLoadingTrack;
+				if( nd.toElement().attribute("type").toInt() == Track::BBTrack ){
+					n += nd.toElement().elementsByTagName("bbtrack").at(0).toElement().firstChildElement().childNodes().count();
```
A bit too long line here.

zonkmachine commented:
> @zonkmachine I fixed code now. Could you test this again?

Crash fixed!

PhysSong commented:
How about conventions?

zonkmachine commented on src/core/Song.cpp:
```diff
@@ -1054,6 +1054,25 @@ void Song::loadProject( const QString & fileName )
+	for( int i=0,n=tclist.count(); i<n; ++i )
+	{
+		QDomNode nd=tclist.at(i).firstChild();
+		while(!nd.isNull()){
```
Here also `{`

zonkmachine commented on src/core/Song.cpp:
```diff
@@ -1054,6 +1054,25 @@ void Song::loadProject( const QString & fileName )
+			if( nd.isElement() && nd.nodeName() == "track" )
+			{
+				++m_nLoadingTrack;
+				if( nd.toElement().attribute("type").toInt() == Track::BBTrack ){
```
And here `{`

PhysSong commented:
@zonkmachine I checked coding conventions, and it seems to be fine.

zonkmachine commented:
I've tested this and it looks just fine to me. I can also confirm the fix in f76cb6fba651d8a22e1955cf85f922aa65642b84 as I managed to replicate the issue. For me opening the demoproject 'DnB' most clearly demonstrated issue  #3611 .

zonkmachine commented:
Merge?

tresf commented:
The code look OK.  If it's all tested, please merge @zonkmachine.  I assume we've tested against a bunch of different project types and that `trackName` logic won't blow up. :)

zonkmachine commented:
> I assume we've tested against a bunch of different project types and that trackName logic won't blow up. :)

The logic doesn't blow up, but the loading time does however. I thought I had tested for that already... :(
Here's a comparison against loading `Skiessi-Onion.mmpz` . I believe lmms-1.0.3 might have loaded faster still. @PhysSong As you can see the loading time increases already with your first commit. This is something people have been worrying about already before this PR.
https://github.com/LMMS/lmms/issues/2029
https://github.com/LMMS/lmms/issues/3312

Do you see any possibilities to trim it down a bit?

```
$ time ../stable-1.1.3/bin/lmms ../data/projects/demos/Skiessi/Skiessi-Onion.mmpz 
real	0m3.499s
user	0m2.296s
sys	0m0.200s

$ time ../stable-1.2.0-RC3/bin/lmms ../data/projects/demos/Skiessi/Skiessi-Onion.mmpz 
real	0m5.042s
user	0m3.864s
sys	0m0.224s

$ time ./lmms-aee0f2b ../data/projects/demos/Skiessi/Skiessi-Onion.mmpz 
real	0m8.044s
user	0m6.656s
sys	0m0.288s

$ time ./lmms-097feb0 ../data/projects/demos/Skiessi/Skiessi-Onion.mmpz
real	0m7.291s
user	0m6.584s
sys	0m0.256s
```

tresf commented:
Does commenting out the [UI call](https://github.com/LMMS/lmms/pull/3672/files#diff-53d10a7a3086e8dea8a3ba36dfa57b1bR125) help?

tresf commented:
I'm also a bit skeptical of [iterating mid-for-loop](https://github.com/LMMS/lmms/pull/3672/files#diff-bc7aaaf624969ba372535c9eab5ed490R1070) although I haven't placed a debugger on it to see if it can cause problems.

zonkmachine commented:
No. The commit just before this PR is:

```
real	0m7.294s
user	0m6.724s
sys	0m0.236s
```
Somethings happened along the way. I'm trying a bisect.

zonkmachine commented:
@PhysSong Sorry, it looks like false alarm. I compiled without `-DCMAKE_BUILD_TYPE=DEBUG` and now get the same time for stable-1.2 and PhysSong:project-load. I thought all my builds were with DEBUG but the one I compared against was not. Without debug the load time I get is about:

```
time ./lmms ../data/projects/demos/Skiessi/Skiessi-Onion.mmpz
real	0m4.552s
user	0m3.868s
sys	0m0.228s
```

This PR is all good.
